### ISSUE TITLE
A new idea for decay as a function of D

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs-benchmark#186]
https://discord.com/channels/368267295601983490/1282005522513530952/1347723544292888586

Right now we estimate S0 like this:
```
            def loss(stability):
                y_pred = self.forgetting_curve(delta_t, stability)
                logloss = sum(
                    -(recall * np.log(y_pred) + (1 - recall) * np.log(1 - y_pred))
                    * count
                )
                l1 = np.abs(stability - init_s0) / 16 if not SECS_IVL else 0
                return logloss + l1

            res = minimize(
                loss,
                x0=init_s0,
                bounds=((S_MIN, INIT_S_MAX),),
                options={"maxiter": int(sum(count))},
            )
```
If we want to make decay depend on D, we could either just assume some fixed value of D...or estimate it from the data!
1) The formula for converting D into decay is very simple. decay=-0.1×D. That's it. When D=1, decay=-0.1, the curve is flat. When D=10, decay=-1.0, the curve is steep.
2) The modified minimize function should look like this:
```
            res = minimize(
                loss,
                x0=[init_s0, init_d0],
                bounds=((S_MIN, INIT_S_MAX), (1, 10)),
                options={"maxiter": int(sum(count))},
            )
```
Now it will fit both S0 and D0 rather than just S0. So now we have a way to estimate D0 (kind of) directly from the data.
3) In pretrain's loss you can do this:
```
            def loss(params):
                stability, difficulty = params[0], params[1]
                y_pred = self.forgetting_curve(delta_t, stability, difficulty)
```
4) In the forgetting curve itself you can do this:
```
    def forgetting_curve(self, t, s, d):
        DECAY = -0.1*d
        FACTOR = 0.9 ** (1 / DECAY) - 1
        return (1 + FACTOR * t / s) ** DECAY
```
5) Then you remove D from the formulas of S and pass D into the forgetting curve instead.

I really hope you will benchmark it.

**Comment by L-M-Sherlock** [comment]
> 5\. Then you remove D from the formulas of S and pass D into the forgetting curve instead.

What about the formula of initial D?

Edit: here is the draft

https://github.com/open-spaced-repetition/fsrs-optimizer/pull/168

**Comment by Expertium** [comment]
There will be no formula for initial D since it's estimated from history
